src/librustc_resolve/macros.rs

```diff
@@ -227,7 +231,11 @@ impl<'a> base::Resolver for Resolver<'a> {
         let name = path[0].name;
         let result = match self.resolve_legacy_scope(&invocation.legacy_scope, name, false) {
             Some(MacroBinding::Legacy(binding)) => Ok(binding.ext.clone()),
-            Some(MacroBinding::Modern(binding)) => Ok(binding.get_macro(self)),
+            Some(MacroBinding::Modern(binding)) => {
+                let ext = binding.get_macro(self);
+                self.err_if_macro_use_proc_macro(name, span, binding, &*ext);
+                Ok(ext)
+            },
```

Comment:
Could you move this [here](https://github.com/abonander/rust/blob/f2619dc9a2fb20a5619dfbd7bb5427d63046dcaa/src/librustc_resolve/macros.rs#L383-L384), i.e. insert between those two lines:
```rust
                (Some(MacroBinding::Modern(binding)), Err(..)) => {
                    self.err_if_macro_use_proc_macro(name, span, binding, &*binding.get(self));
                    continue
                }
```
(would make sense to refactor away the last argument of `err_if_macro_use_proc_macro` then)

With this change, we longer require the user to remove the `#[macro_use]`, only to add a `use` import. Once #39060 lands, they will see unused `#[macro_use]` warnings if they don't remove the `#[macro_use]` and it isn't importing anything else that's being used (e.g. existing custom derives).